xbmc/music/MusicThumbLoader.cpp

```diff
@@ -196,44 +163,58 @@ bool CMusicThumbLoader::FillLibraryArt(CFileItem &item)
   if (tag.GetDatabaseId() > -1 && !tag.GetType().empty())
   {
     m_musicDatabase->Open();
-    std::map<std::string, std::string> artwork;
-    if (m_musicDatabase->GetArtForItem(tag.GetDatabaseId(), tag.GetType(), artwork))
-      item.SetArt(artwork);
-    else if (tag.GetType() == MediaTypeSong)
-    { // no art for the song, try the album
-      ArtCache::const_iterator i = m_albumArt.find(tag.GetAlbumId());
-      if (i == m_albumArt.end())
-      {
-        m_musicDatabase->GetArtForItem(tag.GetAlbumId(), MediaTypeAlbum, artwork);
-        i = m_albumArt.insert(make_pair(tag.GetAlbumId(), artwork)).first;
-      }
-      if (i != m_albumArt.end())
-      {
-        item.AppendArt(i->second, MediaTypeAlbum);
-        for (std::map<std::string, std::string>::const_iterator j = i->second.begin(); j != i->second.end(); ++j)
-          item.SetArtFallback(j->first, "album." + j->first);
-      }
-    }
-    if (tag.GetType() == MediaTypeSong || tag.GetType() == MediaTypeAlbum)
-    { // fanart from the artist
-      std::string fanart = m_musicDatabase->GetArtistArtForItem(tag.GetDatabaseId(), tag.GetType(), "fanart");
-      if (!fanart.empty())
-      {
-        item.SetArt("artist.fanart", fanart);
-        item.SetArtFallback("fanart", "artist.fanart");
-      }
-      else if (tag.GetType() == MediaTypeSong)
+    std::vector<ArtForThumbloader> art;
+    bool artfound;
+    if (tag.GetType() == MediaTypeSong)
+      artfound = m_musicDatabase->GetArtForItem(tag.GetDatabaseId(), tag.GetAlbumId(), -1, true, art);
+    else if (tag.GetType() == MediaTypeAlbum)
+      artfound = m_musicDatabase->GetArtForItem(-1, tag.GetDatabaseId(), -1, true, art);
+    else //Artist
+      artfound = m_musicDatabase->GetArtForItem(-1, -1, tag.GetDatabaseId(), true, art);
+    
+    m_musicDatabase->Close();
+    if (artfound)
```

Comment:
I think this isn't really needed here and at other places, or can be `artfound` false and `art` vector contain elements?

Reply:
If art started with elements then GetArtForItem does not remove them, so using the result flag is more definative.